File: app/src/main/java/com/sylcolabs/weatherpaper/scene/Overlay.kt
package com.sylcolabs.weatherpaper.scene

import android.graphics.Canvas
import android.graphics.Paint
import java.text.Normalizer
import kotlin.math.roundToInt

/** What the home-screen readout shows and where it sits. Persisted in SharedPreferences. */
internal data class OverlayConfig(
    val enabled: Boolean = true,
    val showClock: Boolean = true,
    val showTemp: Boolean = true,
    val showCondition: Boolean = false,
    val showLocation: Boolean = false,
    val x: Float = 0.5f,
    val y: Float = 0.30f,
    val size: Int = 2,
    val clock24: Boolean = true,
) {
    fun hasContent() = enabled && (showClock || showTemp || showCondition || showLocation)
}

/**
 * The home-screen readout, drawn in the bitmap font over whatever the scene is.
 *
 * Never drawn on the lock screen: the artwork beneath is identical either way, so locking reads
 * as the text fading out rather than the whole scene re-laying-out.
 *
 * Colours here are deliberately neutral - a light ink with a dark outline reads over both a
 * bright sky and a dark canopy without knowing anything about the artwork behind it.
 */
internal object Overlay {

    private const val INK = 0xFFECF0EA.toInt()
    private const val OUTLINE = 0xC8060A08.toInt()

    private val paint = Paint().apply {
        isAntiAlias = false
        isDither = false
        style = Paint.Style.FILL
    }

    /** Fold a place name down to what the font can draw. */
    fun normalise(text: String): String =
        Normalizer.normalize(text.uppercase(), Normalizer.Form.NFD).replace(Regex("\\p{Mn}+"), "")

    fun conditionWord(st: SceneState): String = when {
        st.thunder -> "STORM"
        st.precip == Precipitation.DRIZZLE -> "DRIZZLE"
        st.precip == Precipitation.RAIN || st.precip == Precipitation.HEAVY_RAIN -> "RAIN"
        st.precip == Precipitation.SNOW -> "SNOW"
        st.condition == SkyCondition.FOG -> "FOG"
        st.condition == SkyCondition.OVERCAST -> "OVERCAST"
        st.condition == SkyCondition.PARTLY -> "PARTLY"
        else -> "CLEAR"
    }

    fun clockText(st: SceneState, clock24: Boolean): String {
        val h = st.hour.toInt() % 24
        val m = ((st.hour - st.hour.toInt()) * 60f).toInt().coerceIn(0, 59)
        val mm = if (m < 10) "0$m" else "$m"
        if (clock24) return (if (h < 10) "0$h" else "$h") + ":" + mm
        val h12 = if (h % 12 == 0) 12 else h % 12
        return "$h12:$mm"
    }

    private class Line(val text: String, val bump: Int)

    private fun lines(st: SceneState, cfg: OverlayConfig): List<Line> {
        val out = ArrayList<Line>(3)
        if (cfg.showClock) out.add(Line(clockText(st, cfg.clock24), 1))

        val second = StringBuilder()
        if (cfg.showTemp && st.tempC != null) second.append(st.tempC.roundToInt()).append('°')
        if (cfg.showCondition) {
            if (second.isNotEmpty()) second.append(' ')
            second.append(conditionWord(st))
        }
        if (second.isNotEmpty()) out.add(Line(second.toString(), 0))

        if (cfg.showLocation && !st.place.isNullOrBlank()) out.add(Line(normalise(st.place), 0))
        return out
    }

    /**
     * Draw one string, with a 1px dark outline so it stays legible over anything.
     *
     * @param unit size of one font pixel in screen pixels, so the text stays on the same grid as
     *             the artwork rather than being independently scaled.
     */
    fun drawText(canvas: Canvas, text: String, left: Float, top: Float, unit: Float) {
        val fw = PixelFont.WIDTH
        val fh = PixelFont.HEIGHT
        val cols = text.length * PixelFont.ADVANCE
        val mask = BooleanArray(cols * fh)

        for (i in text.indices) {
            val g = PixelFont.glyph(text[i])
            for (gy in 0 until fh) for (gx in 0 until fw) {
                if (PixelFont.ink(g, gx, gy)) mask[gy * cols + i * PixelFont.ADVANCE + gx] = true
            }
        }

        fun cell(cx: Int, cy: Int) {
            val x = left + cx * unit
            val y = top + cy * unit
            canvas.drawRect(x, y, x + unit, y + unit, paint)
        }

        paint.color = OUTLINE
        for (cy in 0 until fh) for (cx in 0 until cols) {
            if (mask[cy * cols + cx]) continue
            var near = false
            outer@ for (dy in -1..1) for (dx in -1..1) {
                val nx = cx + dx; val ny = cy + dy
                if (nx in 0 until cols && ny in 0 until fh && mask[ny * cols + nx]) { near = true; break@outer }
            }
            if (near) cell(cx, cy)
        }
        paint.color = INK
        for (cy in 0 until fh) for (cx in 0 until cols) if (mask[cy * cols + cx]) cell(cx, cy)
    }

    /**
     * @param bounds the artwork's rectangle on screen, which the readout positions itself within.
     * @param unit size of one artwork pixel in screen pixels.
     */
    fun draw(canvas: Canvas, st: SceneState, cfg: OverlayConfig, bounds: android.graphics.RectF, unit: Float) {
        if (!cfg.hasContent()) return
        val ls = lines(st, cfg)
        if (ls.isEmpty()) return

        val size = cfg.size.coerceIn(1, 4)
        val gap = 3f * size * unit

        var total = 0f
        for (l in ls) total += PixelFont.HEIGHT * (size + l.bump) * unit + gap
        total -= gap

        var y = bounds.top + cfg.y.coerceIn(0f, 1f) * bounds.height() - total / 2f
        for (l in ls) {
            val sc = size + l.bump
            val w = PixelFont.width(l.text, sc) * unit
            val x = bounds.left + cfg.x.coerceIn(0f, 1f) * bounds.width() - w / 2f
            drawText(canvas, l.text, x, y, sc * unit)
            y += PixelFont.HEIGHT * sc * unit + gap
        }
    }
}

File: tools/overlay.js
(function (root, factory) {
  if (typeof module === 'object' && module.exports) module.exports = factory(require('./render.js'), require('./scene.js'));
  else root.WeatherPaperOverlay = factory(root.WeatherPaper, root.WeatherPaperScene);
})(typeof self !== 'undefined' ? self : this, function (R, S) {
  'use strict';

  const clamp = R.clamp, mul = R.mul, mix = R.mix;

  /** Fold a place name down to what the 5x7 font can actually draw. */
  function normalise(text) {
    let s = String(text).toUpperCase();
    if (s.normalize) s = s.normalize('NFD').replace(/[̀-ͯ]/g, '');
    return s;
  }

  function glyph(spec, ch) {
    return spec.font.glyphs[ch] || (ch === ' ' ? spec.font.glyphs[' '] : spec.font.glyphs['?']);
  }

  function textWidth(spec, text, scale) {
    const adv = spec.font.w + spec.font.tracking;
    return text.length === 0 ? 0 : (text.length * adv - spec.font.tracking) * scale;
  }

  /**
   * Draw a string with a 1px dark outline. The outline is what keeps the readout legible over
   * both a bright noon sky and a near-black night canopy without needing a panel behind it.
   */
  function drawText(f, spec, text, x, y, scale, ink, outline) {
    const fw = spec.font.w, fh = spec.font.h, adv = fw + spec.font.tracking;
    const cols = text.length * adv;
    const mask = new Uint8Array(cols * fh);

    for (let i = 0; i < text.length; i++) {
      const rows = glyph(spec, text[i]).split('/');
      for (let gy = 0; gy < fh; gy++) {
        for (let gx = 0; gx < fw; gx++) {
          if (rows[gy][gx] === '#') mask[gy * cols + i * adv + gx] = 1;
        }
      }
    }

    function block(cx, cy, colour, alpha) {
      for (let sy = 0; sy < scale; sy++) {
        for (let sx = 0; sx < scale; sx++) f.blend(x + cx * scale + sx, y + cy * scale + sy, colour, alpha);
      }
    }

    for (let cy = 0; cy < fh; cy++) {
      for (let cx = 0; cx < cols; cx++) {
        if (mask[cy * cols + cx]) continue;
        let near = false;
        for (let dy = -1; dy <= 1 && !near; dy++) {
          for (let dx = -1; dx <= 1; dx++) {
            const nx = cx + dx, ny = cy + dy;
            if (nx >= 0 && nx < cols && ny >= 0 && ny < fh && mask[ny * cols + nx]) { near = true; break; }
          }
        }
        if (near) block(cx, cy, outline, 0.78);
      }
    }
    for (let cy = 0; cy < fh; cy++) {
      for (let cx = 0; cx < cols; cx++) if (mask[cy * cols + cx]) block(cx, cy, ink, 1);
    }
  }

  const CONDITION_WORD = {
    none: null, drizzle: 'DRIZZLE', rain: 'RAIN', heavy_rain: 'RAIN', snow: 'SNOW',
  };

  function conditionWord(st) {
    if (st.thunder) return 'STORM';
    const p = CONDITION_WORD[st.precip];
    if (p) return p;
    if (st.condition === 'fog') return 'FOG';
    if (st.condition === 'overcast' || st.cloud > 0.85) return 'OVERCAST';
    if (st.cloud > 0.35) return 'PARTLY';
    return 'CLEAR';
  }

  function formatClock(st) {
    const h = Math.floor(st.hour) % 24;
    const m = Math.floor((st.hour - Math.floor(st.hour)) * 60);
    if (st.clock12) {
      const h12 = h % 12 === 0 ? 12 : h % 12;
      return h12 + ':' + String(m).padStart(2, '0');
    }
    return String(h).padStart(2, '0') + ':' + String(m).padStart(2, '0');
  }

  /** Build the lines the overlay will draw, honouring the user's content toggles. */
  function lines(spec, st, cfg) {
    const out = [];
    if (cfg.showClock) out.push({ text: formatClock(st), scale: 1 });

    const second = [];
    if (cfg.showTemp && st.tempC != null) second.push(Math.round(st.tempC) + '°');
    if (cfg.showCondition) second.push(conditionWord(st));
    if (second.length) out.push({ text: second.join(' '), scale: 0 });

    if (cfg.showLocation && st.place) out.push({ text: normalise(st.place), scale: 0 });
    return out;
  }

  /**
   * Draw the home-screen readout. Never called on the lock screen - the forest beneath is
   * identical, so locking reads as the text fading out rather than the scene re-laying-out.
   */
  function drawOverlay(f, spec, ctx, st, cfg) {
    if (!cfg || cfg.enabled === false) return;
    const ls = lines(spec, st, cfg);
    if (!ls.length) return;

    const size = Math.max(1, Math.round(cfg.size || 2));
    const ink = mix([236, 240, 234], mul([236, 240, 234], ctx.tint), 0.45);
    const outline = [6, 10, 8];
    const gap = Math.round(3 * size);

    let total = 0;
    ls.forEach(function (l) { total += spec.font.h * (size + l.scale) + gap; });
    total -= gap;

    let y = Math.round(clamp(cfg.y, 0, 1) * f.h - total / 2);
    ls.forEach(function (l) {
      const sc = size + l.scale;
      const w = textWidth(spec, l.text, sc);
      const x = Math.round(clamp(cfg.x, 0, 1) * f.w - w / 2);
      drawText(f, spec, l.text, x, y, sc, ink, outline);
      y += spec.font.h * sc + gap;
    });
  }

  return {
    drawOverlay: drawOverlay, drawText: drawText, textWidth: textWidth,
    conditionWord: conditionWord, formatClock: formatClock, normalise: normalise, lines: lines,
  };
});
